### PR TITLE
Draft terrain

### LINKED ISSUE #1662: Make SWIG support Terrain
currently there is no terrain support in swig for C#

### PR DESCRIPTION
**Pull request by natan-fernandes**
fixes #1662

### CODE REVIEW

File: Components/Terrain/include/OgreTerrain.i [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+// #include "OgreUnifiedHighLevelGpuProgram.h"
+%}
+
+%include std_string.i
+%include exception.i 
+%import "Ogre.i"
+
+#define _OgreTerrainExport
+
+%include "OgreTerrainLayerBlendMap.h"
```
Comment (paroj) [inline Components/Terrain/include/OgreTerrain.i:21]: to be sure, you should also `#include` these in the section above.

File: Components/Terrain/include/OgreTerrain.i [hunk 2]
```diff
@@ -0,0 +1,47 @@
+%module(package="Ogre") Terrain
+%{
+/* Includes the header in the wrapper code */
+#include "Ogre.h"
+#include "OgreTerrain.h"
+#include "OgreTerrainAutoUpdateLod.h"
+#include "OgreTerrainGroup.h"
+#include "OgreTerrainPagedWorldSection.h"
```
Comment (paroj) [inline Components/Terrain/include/OgreTerrain.i:8]: for now, you should stay away for the `*Paged*` stuff, as it depends on a separate component and thus complicates stuff.
Paging is not needed for "normal" Terrain usage.

File: Components/Terrain/include/OgreTerrain.i (hunk 1 above)
Comment (paroj) [inline Components/Terrain/include/OgreTerrain.i:21]: you must `%include "OgreTerrain.h"` first, so TerrainLayerBlendMap knows about it. This is what the error:
```
/home/runner/work/ogre/ogre/build/csharp/OgreTerrainCSHARP_wrap.cxx:450:3: error: ‘Terrain’ was not declared in this scope
   Terrain *arg1 = (Terrain *) 0 ;
   ^~~~~~~
/home/runner/work/ogre/ogre/build/csharp/OgreTerrainCSHARP_wrap.cxx:450:3: note: suggested alternative:
In file included from /home/runner/work/ogre/ogre/build/csharp/OgreTerrainCSHARP_wrap.cxx:383:0:
/home/runner/work/ogre/ogre/Components/Terrain/include/OgreTerrain.h:261:30: note:   ‘Ogre::Terrain’
```
means.

Comment (natan-fernandes) [reply Components/Terrain/include/OgreTerrain.i:21]: Should I do a `%include` for every `#include`? I thought it wasn't necessary

Comment (paroj) [reply Components/Terrain/include/OgreTerrain.i:21]: generally, it is not necessary. However %include is not recursive, so you might end up with missing declarations.

**Comment by natan-fernandes** [comment]
Clicked on ready for review by mistake.
I did all the changes you said and I get this error on `OgreTerrain.h`:
`Syntax error - possibly a missing semicolon` at line: 1854

But I don't see any missing semicolon
https://github.com/OGRECave/ogre/blob/master/Components/Terrain/include/OgreTerrain.h#L1854

**Comment by paroj** [comment]
add `#define OGRE_FORCE_INLINE`

**Comment by natan-fernandes** [comment]
Now I get 2 erros: 
```
Macro 'OGRE_FORCE_INLINE' redefined @ OgreTerrain.i
previous definition of 'OGRE_FORCE_INLINE'. @ OgrePlataform.h ln 133
```

**Comment by natan-fernandes** [comment]
Where can I remove the previous `#define` or tell it to aways force?

File: Components/Terrain/include/OgreTerrain.i [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+//#include "OgrePageManager.h"
+//#include "OgrePagedWorld.h"
+//#include "OgreTerrainPaging.h"
+%}
+
+%include std_string.i
+%include exception.i 
+%import "Ogre.i"
+
+#define _OgreTerrainExport
```
Comment (paroj) [inline Components/Terrain/include/OgreTerrain.i:19]:
```suggestion
#define _OgreTerrainExport
#define OGRE_FORCE_INLINE
```
should be here

Comment (natan-fernandes) [reply Components/Terrain/include/OgreTerrain.i:19]: Had the same issue 🙁: 
```
Macro 'OGRE_FORCE_INLINE' redefined.
```

The previous definition is at `OgrePlataform.h`. Somehow the file I have is different from the one on this repository. But here's how it's in mine
```
/* See if we can use __forceinline or if we need to use __inline instead */
#if OGRE_COMPILER_MIN_VERSION(OGRE_COMPILER_MSVC, 1200)
    #define OGRE_FORCE_INLINE __forceinline
#elif OGRE_COMPILER_MIN_VERSION(OGRE_COMPILER_GNUC, 340)
    #define OGRE_FORCE_INLINE inline __attribute__((always_inline))
#else
        #define OGRE_FORCE_INLINE __inline   <----------------------- This is being defined
#endif
```

File: Components/Terrain/include/OgreTerrain.i [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+#include "OgreTerrainGroup.h"
+#include "OgreTerrainLayerBlendMap.h"
+%}
+
+%include std_string.i
+%include exception.i 
+%import "Ogre.i"
+
+#define _OgreTerrainExport
+#define OGRE_FORCE_INLINE
```
Comment (paroj) [inline Components/Terrain/include/OgreTerrain.i:16]: maybe this?
```suggestion
#define __inline
#define __forceinline
```

File: Components/Terrain/include/OgreTerrain.i [hunk 5]
```diff
@@ -0,0 +1,24 @@
+%module(package="Ogre") Terrain
```
Comment (paroj) [inline Components/Terrain/include/OgreTerrain.i:1]:
```suggestion
#ifdef SWIGPYTHON
%module(package="Ogre") Terrain
#else
%module OgreTerrain
#endif
```

**Comment by natan-fernandes** [comment]
Working! 
Now I get 
```
'TerrainLayerDeclaration': undeclared identifier
```
and already tried `#include` and `%include` "OgreTerrainMaterialGenerator.h" but had no success. Is there any header am I missing?

**Comment by paroj** [comment]
you need to add that before `%include "OgreTerrain.h"`. If that gives you other errors you can add `%ignore Ogre::TerrainLayerDeclaration`.

**Comment by natan-fernandes** [comment]
Already tried 
```
%nodefaultctor;
    %include "OgreTerrainGroup.h"
%clearnodefaultctor;
```
but didn't work

**Comment by paroj** [comment]
please update this PR, so I can take a look at the full error log. Probably you should %ignore the fuction associated with MapIterator

**Comment by natan-fernandes** [comment]
If I ignore it will I be able to use it in my C# code?

**Comment by paroj** [comment]
```
%ignore TerrainGroup::getTerrainIterator
```
it will not show up in C#, but you would not be able to use it anyway as the IteratorWrapper does not work with SWIG.

Instead, something like this needs to be done:
https://github.com/OGRECave/ogre/blob/fc789998fa455ea8a4e930c79d40cb930f1f1319/OgreMain/include/OgreSceneNode.h#L233-L245

**Comment by natan-fernandes** [comment]
I need this to do this like Mogre did in the past:
```
foreach (TerrainGroup.TerrainSlot ts in mTerrainGroup.GetTerrainIterator())
{
    ...
}
```

So I need to search for all objects that are `TerrainSlot`? using `getAttachedObjects()`

**Comment by natan-fernandes** [comment]
I will try to do this. 
Also is there a way to create a pointer to a Vector3 like this?
```
Vector3 vAux = new Vector3(1);
Vector3* vPosTerrain = &vAux;
```

It throws `cannot take the address of get the size of or declare a pointer to a managed type(Vector3)`. Is it related to SWIG?

**Comment by paroj** [comment]
assuming the bindings are correct, you never should have to use `Vector3*`

File: Components/Terrain/include/OgreTerrain.i [hunk 6]
```diff
@@ -24,8 +24,12 @@
 %include "OgreTerrainMaterialGenerator.h"
 %include "OgreTerrainLayerBlendMap.h"
 %include "OgreTerrainQuadTreeNode.h"
+
+%template(LayerInstanceList) std::vector<Ogre::Terrain::LayerInstance>;
+%template(Pair) std::pair<bool, Ogre::Vector3>;
```
Comment (natan-fernandes) [inline Components/Terrain/include/OgreTerrain.i:33]: Need to find a better name to Pair, as there are another Pairs in the code that aren't <bool, Vector3>

**Comment by paroj** [comment]
@natan-fernandes are you still working on this?

**Comment by natan-fernandes** [comment]
I got a job so atm I don't have time to work on it but I've made some changes on some cpp files to add `getTerrainSlots()`, I will do a PR tomorrow

**Pull request merged**